cmd/peer-rest-client.go

```diff
@@ -0,0 +1,450 @@
+/*
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/gob"
+	"io"
+	"net/url"
+
+	"github.com/minio/minio/cmd/http"
+	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/cmd/rest"
+	"github.com/minio/minio/pkg/event"
+	xnet "github.com/minio/minio/pkg/net"
+	"github.com/minio/minio/pkg/policy"
+)
+
+// Abstracts a remote disk.
+type peerRESTClient struct {
+	host       *xnet.Host
+	restClient *rest.Client
+	connected  bool
+}
+
+// Returns a peer rest client.
+func (client *peerRESTClient) reConnect() error {
+
+	scheme := "http"
+	if globalIsSSL {
+		scheme = "https"
+	}
+
+	serverURL := &url.URL{
+		Scheme: scheme,
+		Host:   client.host.String(),
+		Path:   peerRESTPath,
+	}
+
+	var tlsConfig *tls.Config
+	if globalIsSSL {
+		tlsConfig = &tls.Config{
+			ServerName: client.host.String(),
+			RootCAs:    globalRootCAs,
+		}
+	}
+
+	restClient, err := rest.NewClient(serverURL, tlsConfig, rest.DefaultRESTTimeout, newAuthToken)
+
+	if err != nil {
+		client.connected = false
+		return err
+	}
+
+	client.restClient = restClient
+	client.connected = true
+	return nil
+}
```

Comment:
Note that restClient is stateless (i.e there is no `connected` state in it)
i.e if you replace this function with:
func (client *peerRESTClient) reConnect() error {
    client.connected = true
}

the net effect will be same.

client.connected will add value only if we implement retry logic with backoff. i.e if it retries to connect more intelligently instead of retrying everytime.

Reply:
If you'd like, you can put this smaller function here and add a comment saying we'll implement retry logic in future and take this PR in now

Reply:
ok